### PR TITLE
Support buffering requests with unknown length (Transfer-Encoding: chunked) in RequestBodyBufferMiddleware

### PR DESCRIPTION
**Pull request by andig**
Implement chunked response handling, following discussion in https://github.com/reactphp/http/issues/232

### CODE REVIEW

**Comment by andig** [comment]
@WyriHaximus I need your help with rebasing this for https://github.com/reactphp/promise-stream/pull/3. In the tests there's currently this for the `RequestBodyBufferMiddlewareTest`:

    public function testExcessiveSizeReturnsError413()
    {
        $stream = new BufferStream(2);
        $stream->write('aa');

        $serverRequest = new ServerRequest(
            'GET',
            'https://example.com/',
            array(),
            $stream
        );

        $buffer = new RequestBodyBufferMiddleware(1);
        $response = $buffer(
            $serverRequest,
            function () {}
        );

        $this->assertSame(413, $response->getStatusCode());
    }

Now what goes in is NOT a `ReadableStreamInterface` and therefore returned early before any buffering in https://github.com/reactphp/http/blob/master/src/Middleware/RequestBodyBufferMiddleware.php#L43. 

Is that really desired, i.e. should it be possible to construct `ServerRequest` with a body that doesn't implement Psr7 semantics (and therefore actually is a `HttpBodyStream`? Hidden bug or feature that I fail to grasp?

**Comment by clue** [comment]
@andig Have you seen the documentation for the [`RequestBodyBufferMiddleware`](https://github.com/reactphp/http#requestbodybuffermiddleware)? I think that this middleware should support all valid requests. In a normal stack this will be the incoming streaming request, but another prior middleware handler may also expose a buffered request to this middleware handler.

The following (completely contrived) example should allow request up to 8 MiB:
```
$middleware = new MiddlewareRunner([
    new RequestBodyBufferMiddleware(16 * 1024 * 1024), // 16 MiB
    new RequestBodyBufferMiddleware(8 * 1024 * 1024), // 8 MiB
});
```

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 1]
```diff
@@ -29,27 +30,23 @@ public function __construct($sizeLimit = null)
...
     {
-        $size = $request->getBody()->getSize();
-
-        if ($size === null) {
-            return new Response(411, array('Content-Type' => 'text/plain'), 'No Content-Length header given');
-        }
-
-        if ($size > $this->sizeLimit) {
-            return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
-        }
```
Comment (clue) [inline src/Middleware/RequestBodyBufferMiddleware.php]: We should probably keep this check: it makes sense to fail early in case the client is trying to upload a gigabyte file over a slow uplink :-)

Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Fixed. Please forgive me for force-pushing. I didn't realize you were already reviewing.

**Comment by andig** [comment]
Small side note: I've also changed mentions of `g7` and `Psr7Implemention` to `Psr7` for sake of coding consistency.

**Comment by andig** [comment]
@clue 

> Have you seen the documentation for the `RequestBodyBufferMiddleware`? I think that this middleware should support all valid requests.

Reading this:

> This can be useful if full PSR-7 compatibility is needed for the request handler and the default streaming request body handling is not needed.

I understand the behaviour. It's still totally confusing though that- if you use the `RequestBodyBufferMiddleware` and send in a react stream with valid body it can be of any size and will *not* be buffered or rejected.

Shouldn't at least the `RequestBodyBufferMiddleware` reject non-Psr7 bodies as they are used in the tests?

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 1 above)
Comment (clue) [reply src/Middleware/RequestBodyBufferMiddleware.php]: No worries, happy to see you're faster with coding than I am with reviewing :shipit:

**Comment by clue** [comment]
I'm not sure I follow tbh. Our middleware implementation works around the concept of PSR-7 requests coming in and PSR-7 responses coming out (plus wrapped in promises). This means that there's no such thing as a "non-Psr7 body".

What's special about this implementation is that bodies can be in a streaming state when a `ReadableStreamInterface` is attached. This allows us to transport bodies of arbitrary sizes without having to store huge bodies in memory.

Does this help?

**Comment by andig** [comment]
> I'm not sure I follow tbh.

Never mind. I've just noticed that I had to rewrite https://github.com/reactphp/http/pull/235/files#diff-5ff94d7c5e50a34ed39b11c33717c0b9R68 to use a different stream type. I'm probably making this too complicated ;)

**Comment by WyriHaximus** [comment]
@andig Seems you're getting there, keep up the good work :+1:

File: src/HttpBodyStream.php [hunk 2]
```diff
@@ -25,7 +25,7 @@ class HttpBodyStream extends EventEmitter implements StreamInterface, ReadableSt
      * @param ReadableStreamInterface $input - Stream data from $stream as a body of a PSR-7 object4
      * @param int|null $size - size of the data body
      */
-    public function __construct(ReadableStreamInterface $input, $size)
+    public function __construct(ReadableStreamInterface $input, $size = null)
```
[the remark this answers is not in the document]
Comment (andig) [reply src/HttpBodyStream.php]: Imho this change makes sense. It's aligned with the comment now, the constructor handles null and it prevents explicitly passing null for bodies of unknown length?

File: src/RequestHeaderParser.php [hunk 3]
```diff
@@ -92,7 +92,7 @@ private function parseRequest($headers)
         }
 
         // parse request headers into obj implementing RequestInterface
-        $request = g7\parse_request($headers);
+        $request = Psr7\parse_request($headers);
```
[the remark this answers is not in the document]
Comment (andig) [reply src/RequestHeaderParser.php]: I wanted to have Psr7 consistently named in the codebase (not Psr7, g7, and PsrImplementation). Could move to separate PR?

File: src/Response.php [hunk 4]
```diff
@@ -21,7 +21,7 @@ public function __construct(
         $version = '1.1',
         $reason = null
     ) {
-        if ($body instanceof ReadableStreamInterface) {
+        if ($body instanceof ReadableStreamInterface &! $body instanceof HttpBodyStream) {
```
[the remark this answers is not in the document]
Comment (andig) [reply src/Response.php]: Artifact removed.

File: src/Server.php [hunk 5]
```diff
@@ -3,13 +3,12 @@
 namespace React\Http;
```
[the remark this answers is not in the document]
Comment (andig) [reply src/Server.php]: Will open separate PR for Psr7 but imho at least this change https://github.com/reactphp/http/pull/235/files#diff-c611f5024b5b13f69f779505d1c6735eR170 makes sense.

Do you mind the reordering of the use clauses by library?

File: src/HttpBodyStream.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (clue) [reply src/HttpBodyStream.php]: I feel that this adds noise and adds no value and as part of this PR, maybe suggest this in a separate PR? :+1: 

Despite this, I don't see what value this would provide, as this is an internal class only and this change basically means you're allowed th instantiate the class without considering what its "size" would be. If you feel this is relevant, I encourage you to file a PR so we can discuss this further :+1:

**Comment by WyriHaximus** [comment]
@andig :tada: https://github.com/reactphp/promise-stream/releases/tag/v0.1.2 :tada:

**Comment by WyriHaximus** [comment]
@andig updated this PR's title to something that reflects what the current code is doing

**Comment by andig** [comment]
I've added one more commit to test for 413 in case on known but excessive content length. Still need to have another look at the README.

**Comment by WyriHaximus** [comment]
@andig could you update to `^1.0 || ^0.1.2` since `1.0` has been release earlier today? :tada:

File: README.md [hunk 6]
```diff
@@ -697,16 +697,17 @@ Any incoming request that has a request body that exceeds this limit will be
...
-transfer encoding for outgoing responses is not affected.
-It is recommended to define a `Content-Length` header instead.
-Note that this does not affect normal requests without a request body
-(such as a simple `GET` request).
+The `RequestBodyBufferMiddleware` will buffer requests with bodies of known size 
+(i.e. with `Content-Length` header specified) as well as requests with bodies of 
+unknown size (i.e. with `Transfer-Encoding: chunked` header). To simplify request 
+handling for downstream middlewares or applications buffered `chunked` requests will 
+have the `Transfer-Encoding` header removed and a `Content-Length` header with the 
+actual request body size received added instead.
```
Comment (clue) [inline README.md]: README updates are much appreciated :+1: However, I feel this is a bit misleading here, as this makes it sounds like this is what this middleware handler does, whereas this is actually handled by the `Server` itself. Also, the word "middlewares" doesn't exist (think "softwares"), I usually use "middleware handlers" instead :-) Can you rephrase this? :+1:

File: README.md [hunk 7]
```diff
@@ -697,16 +697,17 @@ Any incoming request that has a request body that exceeds this limit will be
...
+
+Note that this only affects incoming requests, the chunked transfer encoding for 
+outgoing responses is not affected.
```
Comment (clue) [inline README.md]: This sentence made much more sense for the old behavior where these requests were rejected. Does it make sense to you to remove this sentence?

File: README.md (hunk 6 above)
Comment (andig) [reply README.md]: I think it has actually become a duplicate since the response section already describes the behaviour. Removed.

File: README.md (hunk 7 above)
Comment (andig) [reply README.md]: Yep, removed.

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 8]
```diff
@@ -29,27 +30,30 @@ public function __construct($sizeLimit = null)
...
+        $body = $request->getBody();
 
-        if ($size > $this->sizeLimit) {
+        // request body of known size exceeding limit
+        if ($body->getSize() > $this->sizeLimit) {
             return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
         }
 
-        $body = $request->getBody();
         if (!$body instanceof ReadableStreamInterface) {
```
Comment (andig) [inline src/Middleware/RequestBodyBufferMiddleware.php]: Shouldn't this check be moved up as this type of stream won't be buffered anyway? Why complain about size in this case (shouldn't happen when using the server at all, might as well just throw)?

Comment (clue) [reply src/Middleware/RequestBodyBufferMiddleware.php]: IMO it makes sense to keep this like this in (the rare) case this middleware is used behind another buffering middleware handler, see also https://github.com/reactphp/http/pull/235#issuecomment-336528451 :+1:

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 9]
```diff
@@ -29,27 +30,28 @@ public function __construct($sizeLimit = null)
...
+        return Stream\buffer($body, $this->sizeLimit)->then(function ($buffer) use ($request, $stack) {
             $stream = new BufferStream(strlen($buffer));
             $stream->write($buffer);
             $request = $request->withBody($stream);
 
             return $stack($request);
+        }, function($error) {
+            if ($error instanceof OverflowException) {
+                return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
+            }
```
[the remark this answers is not in the document]
Comment (clue) [reply src/Middleware/RequestBodyBufferMiddleware.php]: I think this is a relevant point, let's postpone this discussion for now and address this in a follow-up ticket :shipit: :+1: 

See for example https://stackoverflow.com/questions/14250991/is-it-acceptable-for-a-server-to-send-a-http-response-before-the-entire-request

[the remark this answers is not in the document]
Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Understood. Looking forward to your insights once this is in.

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 8 above)
Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: @clue What I mean is that- if not a `ReadableStreamInterface`, which means we won't buffer it, we're still returning 413 although the middleware is _not_ responsible for handling this request as it is right now.

My suggestion would be to move the `ReadableStreamInterface` to the Beginning of the method?

Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: ping @clue could you check my question above? 
What's missing to make make progress with this PR (and 0.8 in general)?

Comment (clue) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Well, technically buffering and limiting sizes are in fact independent responsibilities. For practical reasons however, I think that it makes sense to bring them close together because buffering without applying a limit is actually harmful (potential for trivial DOS otherwise). If we accept the fact that checking size limits here, then I think it makes sense to also support buffered requests here. This means that I think it makes sense to also reject buffered requests that exceed the given size limits.

Accordingly, I think this code LGTM :+1: What do you think about this?

File: src/Middleware/RequestBodyBufferMiddleware.php [hunk 10]
```diff
@@ -29,27 +30,30 @@ public function __construct($sizeLimit = null)
...
             $request = $request->withBody($stream);
 
             return $stack($request);
+        }, function($error) {
+            // request body of unknown size exceeding limit during buffering
+            if ($error instanceof OverflowException) {
+                return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
+            }
+
+            return $error;
```
Comment (clue) [inline src/Middleware/RequestBodyBufferMiddleware.php]: This probably shouldn't `return` an `Exception` here :-) Can you fix this and add a test?

Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: @clue good catch! Undecided what to return though- right now it's an HTTP 400 with the Exception message as text. Do you have a better suggestion?

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 8 above)
Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: > What do you think about this?

@clue sratch head. I'm not getting my point across. We're only buffering and limiting PSR streams. We are however *still* limiting react-only streams. 

This is imho wrong- as the buffering middleware is not responsible for these requests. I'm proposing-  to be explicit- to change order of things like this:

    public function __invoke(ServerRequestInterface $request, $stack)
    {
        $body = $request->getBody();

        if (!$body instanceof ReadableStreamInterface) {
            return $stack($request);
        }

        // request body of known size exceeding limit
        if ($body->getSize() > $this->sizeLimit) {
            return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
        }

Comment (clue) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Yes, it looks like we're talking past one another :man_shrugging: :-)

The `getBody()` method is guaranteed to *always* return an instance of `Psr\Message\StreamInterface`. This very instance *may* also implement the `React\Stream\ReadableStreamInterface` ("streaming state"), but this is not guaranteed (already in "buffered state").

When this middleware handler is used as documented, then the body is always in a "streaming state". If one were to use this middleware handler twice in the same stack (conceived example), then the first will see a body in "streaming state" while the second will see it in "buffered state".

If this conceived(!) second middleware handler uses a smaller size limit (which I would expect to be more of a programming error than intentional), then IMO the second middleware handler should still reject the message in my opinion.

If I understand your above suggestion correctly, then it looks like this would actually allow the body to pass through without applying the second limit at all.

I think it makes sense to add a test for this behavior to ensure that this is properly covered. Can you add a test for this or would you rather want me to add this to this PR? :+1:

File: src/Middleware/RequestBodyBufferMiddleware.php (hunk 10 above)
Comment (andig) [reply src/Middleware/RequestBodyBufferMiddleware.php]: Done. Do you think we need to guard for error actually being an exception and not something else?

**Comment by WyriHaximus** [comment]
What is the status here @clue and @andig?

**Comment by andig** [comment]
Needs review of the last 2 commits
